### PR TITLE
feat(source-gcs): added oauth flow

### PR DESCRIPTION
**Pull request by darynaishchenko**
## What
Add supporting OAuth
Additionally: 
resolves: https://github.com/airbytehq/airbyte-internal-issues/issues/9745 (remove discriminator fields)

## How
Extended spec with advanced_auth, added credentials object with config migration that adds deprecated `service_account` field into credentials object.
Added and updated unit tests.

PR for the platform: https://github.com/airbytehq/airbyte-platform-internal/pull/13923

## Review guide
1. `airbyte-integrations/connectors/source-gcs/source_gcs/config.py`
2. `airbyte-integrations/connectors/source-gcs/source_gcs/source.py`
3. `airbyte-integrations/connectors/source-gcs/source_gcs/stream_reader.py`
4. other files

## Can this PR be safely reverted and rolled back?

- [x] YES 💚
- [ ] NO ❌

### CODE REVIEW

File: airbyte-integrations/connectors/source-gcs/integration_tests/spec.json [hunk 1]
```diff
@@ -437,12 +437,83 @@
...
-        "title": "Service Account Information",
-        "description": "Enter your Google Cloud <a href=\"https://cloud.google.com/iam/docs/creating-managing-service-account-keys#creating_service_account_keys\">service account key</a> in JSON format",
-        "airbyte_secret": true,
+      "credentials": {
+        "title": "Authentication",
+        "description": "Credentials for connecting to the Google Cloud Storage API",
+        "type": "object",
         "order": 0,
-        "type": "string"
+        "discriminator": {
```
Comment (girarda) [inline airbyte-integrations/connectors/source-gcs/integration_tests/spec.json:445]: I expect the discriminator fields to trip up the Terraform provider. Maybe we need a method similar to this one where we remove them from the schemas https://github.com/airbytehq/airbyte/blob/fc12432305da7725379f967a64d0b0f011e02d78/airbyte-cdk/python/airbyte_cdk/sources/file_based/config/abstract_file_based_spec.py#L58-L60

File: docs/integrations/sources/gcs.md [hunk 2]
```diff
@@ -151,40 +151,41 @@ Leaving this field blank (default option) will disallow escaping.
...
-| 0.3.5 | 2024-01-30 | [34661](https://github.com/airbytehq/airbyte/pull/34661) | Pin CDK version until upgrade for compatibility with the Concurrent CDK |
-| 0.3.4 | 2024-01-11 | [34158](https://github.com/airbytehq/airbyte/pull/34158) | Fix issue in stream reader for document file type parser |
-| 0.3.3 | 2023-12-06 | [33187](https://github.com/airbytehq/airbyte/pull/33187) | Bump CDK version to hide source-defined primary key |
-| 0.3.2 | 2023-11-16 | [32608](https://github.com/airbytehq/airbyte/pull/32608) | Improve document file type parser |
-| 0.3.1 | 2023-11-13 | [32357](https://github.com/airbytehq/airbyte/pull/32357) | Improve spec schema |
-| 0.3.0 | 2023-10-11 | [31212](https://github.com/airbytehq/airbyte/pull/31212) | Migrated to file based CDK |
-| 0.2.0 | 2023-06-26 | [27725](https://github.com/airbytehq/airbyte/pull/27725) | License Update: Elv2 |
-| 0.1.0 | 2023-02-16 | [23186](https://github.com/airbytehq/airbyte/pull/23186) | New Source: GCS |
+| 0.7.0   | 2024-09-13 | [45414](https://github.com/airbytehq/airbyte/pull/45414) | Add OAuth                                                               |
```
Comment (girarda) [inline docs/integrations/sources/gcs.md:154]:
```suggestion
| 0.7.0   | 2024-09-13 | [45414](https://github.com/airbytehq/airbyte/pull/45414) | Add support for OAuth authentication                                                               |
```

File: airbyte-integrations/connectors/source-gcs/integration_tests/spec.json (hunk 1 above)
Comment (darynaishchenko) [reply airbyte-integrations/connectors/source-gcs/integration_tests/spec.json]: removed `def remove_discriminator` from Config class, so both discriminators are being removed from spec

File: airbyte-integrations/connectors/source-gcs/source_gcs/config_migrations.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        migrated_config = cls.transform(config)
+
+        source.write_config(migrated_config, config_path)
+
+        return migrated_config
+
+    @classmethod
+    def emit_control_message(cls, migrated_config: Mapping[str, Any]) -> None:
+        # add the Airbyte Control Message to message repo
+        cls.message_repository.emit_message(create_connector_config_control_message(migrated_config))
```
[the remark this answers is not in the document]
Comment (darynaishchenko) [reply airbyte-integrations/connectors/source-gcs/source_gcs/config_migrations.py:36]: Printing the message directly should lead to the same result as via repository. But using `message_repository` is a common way to emit control messages for most of config migrations, so I prefer to use it.

**Review by girarda — APPROVED** [review]
PR looks good! @darynaishchenko just to confirm - we need to update our app's scopes to be able to test the new flow, correct?

**Comment by darynaishchenko** [comment]
> PR looks good! @darynaishchenko just to confirm - we need to update our app's scopes to be able to test the new flow, correct?

We need:
1. merge [pr](https://github.com/airbytehq/airbyte-platform-internal/pull/13923) for platform
2. make a dev image of the source based changes in this pr
3. add client id/secret (for example from other google app) to platform

Ask google support to test it for verification. After verification we can add new scopes to our app. cc: @DanyloGL

**Review by lazebnyi — APPROVED** [review]

**Pull request merged**